server/libs/storage/src/storage.service.ts

```diff
@@ -0,0 +1,110 @@
+import { APP_UPLOAD_LOCATION } from '@app/common';
+import { AssetEntity } from '@app/database/entities/asset.entity';
+import { ImmichConfigService } from '@app/immich-config';
+import { Injectable, Logger } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { FileSystemStorageService } from './storage.service.filesystem';
+import handlebar from 'handlebars';
+import * as luxon from 'luxon';
+import sanitize from 'sanitize-filename';
+import path from 'node:path';
+import { constants } from 'node:fs';
+import fsPromise from 'fs/promises';
+import mv from 'mv';
+
+@Injectable()
+export class StorageService {
+  readonly log = new Logger(StorageService.name);
+
+  constructor(
+    @InjectRepository(AssetEntity)
+    private assetRepository: Repository<AssetEntity>,
+
+    private fileSystemStorageService: FileSystemStorageService,
+
+    private immichConfigService: ImmichConfigService,
+  ) {}
+
+  public async moveFile(asset: AssetEntity, filename: string): Promise<AssetEntity> {
+    try {
+      const parsedPath = await this.buildPath(asset, filename);
+      let duplicateCount = 1;
+      let qualifiedPath = path.join(APP_UPLOAD_LOCATION, asset.userId, parsedPath);
+
+      while (await this.checkFileExist(qualifiedPath)) {
+        const newPath = await this.buildPath(asset, filename + `_${duplicateCount}`);
```

Comment:
This logic has some problems with it - specifically if the {{filename}} isn't used in the template, so I think this needs to be refactored to append duplicate count + the `.ext` after getting back the merged template.